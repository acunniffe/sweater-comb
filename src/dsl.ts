import {
  ApiCheckDsl,
  EntityRule,
  Result,
  DocsLinkHelper,
  newDocsLinkHelper,
  runCheck,
} from "@useoptic/api-checks";
import {
  ConceptualLocation,
  OpenApiHeaderFact,
  OpenApiKind,
  OpenApiOperationFact,
  IChange,
  IFact,
  OpenApiFieldFact,
  jsonPointerHelper,
  ILocation,
  OpenAPIV3,
} from "@useoptic/openapi-utilities";
import { genericEntityRuleImpl } from "@useoptic/api-checks/build/sdk/generic-entity-rule-impl";
import { ShouldOrMust } from "@useoptic/api-checks/build/sdk/types";
import { OpenApiResponseFact } from "@useoptic/openapi-utilities/build/openapi3/implementations/openapi3/openapi-traverser";

type SnykStablity = "wip" | "experimental" | "beta" | "ga";
type DateString = string; // YYYY-mm-dd
type ResourceName = string;

export interface SynkApiCheckContext {
  // Vervet provides context about the change itself. Since
  // Optic is analyzing two OpenAPI spec files, we need to tell it
  // when the change is supposed to happen, and the resource/version info
  // determined by the file's location in a directory structure.
  changeDate: DateString; // when the change did (or would, if proposed) occur
  changeResource: ResourceName; // the spec resource being changed
  changeVersion: {
    // the spec version being changed
    date: DateString;
    stability: SnykStablity;
  };

  // Vervet provides a mapping that indicates the resource version deprecation.
  // It has this information because it processes the entire source tree of
  // spec files.
  resourceVersions: {
    [ResourceName: string]: {
      // changeResource used to match this
      [DateString: string]: {
        // changeVersion.date used to match this
        [SnykStablity: string]: {
          // changeVersion.stability matches this
          deprecatedBy: {
            // the spec version that deprecates this one (if any) or null
            date: DateString;
            stability: SnykStablity; // could be higher stability than changed!
          } | null;
        };
      };
    };
  };
}

export interface SnykEntityRule<T, A>
  extends EntityRule<T, ConceptualLocation, SynkApiCheckContext, A> {}

export class SnykApiCheckDsl implements ApiCheckDsl {
  private checks: Promise<Result>[] = [];

  constructor(
    private nextFacts: IFact<any>[],
    private changelog: IChange<any>[],
    private nextJsonLike: OpenAPIV3.Document,
    private providedContext: SynkApiCheckContext
  ) {}

  checkPromises() {
    return this.checks;
  }

  getContext(location: ILocation): ConceptualLocation & SynkApiCheckContext {
    return {
      ...location.conceptualLocation,
      ...this.providedContext,
    };
  }

  get operations() {
    return genericEntityRuleImpl<
      OpenApiOperationFact,
      ConceptualLocation,
      SynkApiCheckContext,
      OpenAPIV3.OperationObject
    >(
      OpenApiKind.Operation,
      this.changelog,
      this.nextFacts,
      (opFact) => `${opFact.method.toUpperCase()} ${opFact.pathPattern}`,
      (location) => this.getContext(location),
      (...items) => this.checks.push(...items),
      (pointer: string) => jsonPointerHelper.get(this.nextJsonLike, pointer)
    );
  }

  get context() {
    const change: IChange<SnykApiCheckDsl> = {
      location: {
        conceptualLocation: { path: "This Specification", method: "" },
        jsonPath: "/",
        conceptualPath: [],
        kind: "API",
      },
    };

    const value: ShouldOrMust<
      (
        context: SynkApiCheckContext,
        docs: DocsLinkHelper
      ) => Promise<void> | void
    > = {
      must: (statement, handler) => {
        const docsHelper = newDocsLinkHelper();
        return runCheck(
          change,
          docsHelper,
          "this specification: ",
          statement,
          true,
          () => handler(this.providedContext, docsHelper)
        );
      },
      should: (statement, handler) => {
        const docsHelper = newDocsLinkHelper();
        return runCheck(
          change,
          docsHelper,
          "this specification: ",
          statement,
          false,
          () => handler(this.providedContext, docsHelper)
        );
      },
    };

    return value;
  }

  get responses() {
    const dsl = this;

    return {
      ...genericEntityRuleImpl<
        OpenApiResponseFact,
        ConceptualLocation,
        SynkApiCheckContext,
        OpenAPIV3.ResponsesObject
      >(
        OpenApiKind.Response,
        dsl.changelog,
        dsl.nextFacts,
        (response) => `response ${response.statusCode}`,
        (location) => dsl.getContext(location),
        (...items) => dsl.checks.push(...items),
        (pointer: string) => jsonPointerHelper.get(dsl.nextJsonLike, pointer)
      ),
      headers: genericEntityRuleImpl<
        OpenApiHeaderFact,
        ConceptualLocation,
        SynkApiCheckContext,
        OpenAPIV3.HeaderObject
      >(
        OpenApiKind.ResponseHeader,
        dsl.changelog,
        dsl.nextFacts,
        (header) => `response header ${header.name}`,
        (location) => dsl.getContext(location),
        (...items) => dsl.checks.push(...items),
        (pointer: string) => jsonPointerHelper.get(dsl.nextJsonLike, pointer)
      ),
    };
  }

  get checkApiContext(): ShouldOrMust<
    (context: SynkApiCheckContext, docs: DocsLinkHelper) => void
  > {
    const contextChangedHandler: (must: boolean) => ContextChangedRule["must"] =
      (must: boolean) => {
        return (statement, handler) => {
          const docsHelper = newDocsLinkHelper();
          const syntheticChange = {
            added: this.providedContext,
            location: {
              jsonPath: "/",
              conceptualPath: [],
              conceptualLocation: {
                path: "Entire Resource",
                method: "",
              },
              kind: "ContextRule",
            },
          };
          return runCheck(
            syntheticChange,
            docsHelper,
            "api lifeycle: ",
            statement,
            must,
            () => handler(this.providedContext, docsHelper)
          );
        };
      };

    return {
      must: contextChangedHandler(true),
      should: contextChangedHandler(false),
    };
  }

  get bodyProperties(): SnykEntityRule<
    OpenApiFieldFact,
    OpenAPIV3.SchemaObject
  > {
    const dsl = this;
    return genericEntityRuleImpl<
      OpenApiFieldFact,
      ConceptualLocation,
      SynkApiCheckContext,
      OpenAPIV3.SchemaObject
    >(
      OpenApiKind.Field,
      dsl.changelog,
      dsl.nextFacts,
      (field) => `field ${field.key}`,
      (location) => dsl.getContext(location),
      (...items) => dsl.checks.push(...items),
      (pointer: string) => jsonPointerHelper.get(dsl.nextJsonLike, pointer)
    );
  }
}

type ContextChangedRule = ShouldOrMust<
  (context: SynkApiCheckContext, docs: DocsLinkHelper) => void
>;
