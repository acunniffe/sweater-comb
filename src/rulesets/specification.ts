import { SnykApiCheckDsl } from "../dsl";
import { expect } from "chai";
import { pascalCase } from "change-case";
import { links } from "../docs";

export const rules = {
  componentNameCase: ({ specification }: SnykApiCheckDsl) => {
    specification.requirement.must(
      "use pascal case for component names",
      (spec, context, docs) => {
        docs.includeDocsLink(links.standards.referencedEntities);
        const componentTypes = Object.keys(spec.components || {});
        for (const componentType of componentTypes) {
          const componentNames = Object.keys(
            spec.components?.[componentType] || {},
          );
          for (const componentName of componentNames) {
            expect(pascalCase(componentName)).to.equal(componentName);
          }
        }
      },
    );
  },
  listOpenApiVersions: ({ specification }: SnykApiCheckDsl) => {
    specification.requirement.must(
      "list the available versioned OpenAPI specifications",
      (spec, context, docs) => {
        docs.includeDocsLink(links.standards.openApiVersions);
        if (spec["x-snyk-api-stability"] === undefined) {
          // Only applicable to compiled OAS; resource versions do not need to declare this
          const pathUrls = Object.keys(spec.paths);
          expect(pathUrls).to.include("/openapi");
        }
      },
    );
  },
  getOpenApiVersions: ({ specification }: SnykApiCheckDsl) => {
    specification.requirement.must(
      "provide versioned OpenAPI specifications",
      (spec, context, docs) => {
        docs.includeDocsLink(links.standards.openApiVersions);
        if (spec["x-snyk-api-stability"] === undefined) {
          // Only applicable to compiled OAS; resource versions do not need to declare this
          const pathUrls = Object.keys(spec.paths);
          expect(pathUrls).to.include("/openapi/{version}");
        }
      },
    );
  },
  orgOrGroupTenant: ({ specification }: SnykApiCheckDsl) => {
    specification.requirement.must(
      "have an org or group tenant",
      (spec, context, docs) => {
        docs.includeDocsLink(links.standards.orgAndGroupTenantResources);
        docs.becomesEffectiveOn(new Date("2021-07-01"));
        const untenantedUrls = Object.keys(spec.paths).filter(
          (url) =>
            !url.startsWith("/orgs/{org_id}") &&
            !url.startsWith("/groups/{group_id}"),
        );
        expect(
          untenantedUrls,
          `expected support for org or group tenant`,
        ).to.have.lengthOf(0);
      },
    );
  },
  tags: ({ specification }: SnykApiCheckDsl) => {
    specification.requirement.must(
      "have name and description for tags",
      (spec, context, docs) => {
        docs.includeDocsLink(links.standards.tags);
        const tags = spec.tags || [];
        for (const tag of tags) {
          expect(tag).to.have.property("name");
          expect(tag).to.have.property("description");
        }
      },
    );
  },
};
