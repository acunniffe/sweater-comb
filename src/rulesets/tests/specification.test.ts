import { rules } from "../specification";
import { SynkApiCheckContext } from "../../dsl";

import { createSnykTestFixture } from "./fixtures";
const { compare } = createSnykTestFixture();

const emptyContext: SynkApiCheckContext = {
  changeDate: "2021-10-10",
  changeResource: "Example",
  changeVersion: {
    date: "2021-10-10",
    stability: "ga",
  },
  resourceVersions: {},
};
