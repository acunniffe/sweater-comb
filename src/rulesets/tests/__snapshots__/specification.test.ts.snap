// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`orgOrGroupTenant fails with both valid and invalid group tenants 1`] = `
Object {
  "base": Object {
    "info": Object {
      "title": "OpenAPI",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {},
  },
  "changelog": Array [],
  "next": Object {
    "info": Object {
      "title": "OpenAPI",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/bad-tenant": Object {},
      "/groups/{group_id}/thing": Object {},
    },
  },
  "results": Array [
    Object {
      "change": Object {
        "location": Object {
          "conceptualLocation": Object {
            "method": "",
            "path": "This Specification",
          },
          "conceptualPath": Array [],
          "jsonPath": "/",
          "kind": "API",
        },
      },
      "condition": "have an org or group tenant",
      "docsLink": "https://github.com/snyk/sweater-comb/blob/main/docs/standards.md#organization-and-group-tenants-for-resources",
      "effectiveOnDate": 2021-07-01T00:00:00.000Z,
      "error": "expected support for org or group tenant: expected [ '/groups/{group_id}/thing' ] to have a length of 2 but got 1",
      "isMust": true,
      "isShould": false,
      "passed": false,
      "where": "this specification: ",
    },
  ],
}
`;

exports[`orgOrGroupTenant fails with both valid and invalid org tenants 1`] = `
Object {
  "base": Object {
    "info": Object {
      "title": "OpenAPI",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {},
  },
  "changelog": Array [],
  "next": Object {
    "info": Object {
      "title": "OpenAPI",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/bad-tenant": Object {},
      "/orgs/{org_id}/thing": Object {},
    },
  },
  "results": Array [
    Object {
      "change": Object {
        "location": Object {
          "conceptualLocation": Object {
            "method": "",
            "path": "This Specification",
          },
          "conceptualPath": Array [],
          "jsonPath": "/",
          "kind": "API",
        },
      },
      "condition": "have an org or group tenant",
      "docsLink": "https://github.com/snyk/sweater-comb/blob/main/docs/standards.md#organization-and-group-tenants-for-resources",
      "effectiveOnDate": 2021-07-01T00:00:00.000Z,
      "error": "expected support for org or group tenant: expected [ '/orgs/{org_id}/thing' ] to have a length of 2 but got 1",
      "isMust": true,
      "isShould": false,
      "passed": false,
      "where": "this specification: ",
    },
  ],
}
`;

exports[`orgOrGroupTenant path '' is valid: false 1`] = `
Object {
  "base": Object {
    "info": Object {
      "title": "OpenAPI",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {},
  },
  "changelog": Array [],
  "next": Object {
    "info": Object {
      "title": "OpenAPI",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "": Object {},
    },
  },
  "results": Array [
    Object {
      "change": Object {
        "location": Object {
          "conceptualLocation": Object {
            "method": "",
            "path": "This Specification",
          },
          "conceptualPath": Array [],
          "jsonPath": "/",
          "kind": "API",
        },
      },
      "condition": "have an org or group tenant",
      "docsLink": "https://github.com/snyk/sweater-comb/blob/main/docs/standards.md#organization-and-group-tenants-for-resources",
      "effectiveOnDate": 2021-07-01T00:00:00.000Z,
      "error": "expected support for org or group tenant: expected [] to have a length of 1 but got +0",
      "isMust": true,
      "isShould": false,
      "passed": false,
      "where": "this specification: ",
    },
  ],
}
`;

exports[`orgOrGroupTenant path '/group/{group_id}' is valid: false 1`] = `
Object {
  "base": Object {
    "info": Object {
      "title": "OpenAPI",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {},
  },
  "changelog": Array [],
  "next": Object {
    "info": Object {
      "title": "OpenAPI",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/group/{group_id}": Object {},
    },
  },
  "results": Array [
    Object {
      "change": Object {
        "location": Object {
          "conceptualLocation": Object {
            "method": "",
            "path": "This Specification",
          },
          "conceptualPath": Array [],
          "jsonPath": "/",
          "kind": "API",
        },
      },
      "condition": "have an org or group tenant",
      "docsLink": "https://github.com/snyk/sweater-comb/blob/main/docs/standards.md#organization-and-group-tenants-for-resources",
      "effectiveOnDate": 2021-07-01T00:00:00.000Z,
      "error": "expected support for org or group tenant: expected [] to have a length of 1 but got +0",
      "isMust": true,
      "isShould": false,
      "passed": false,
      "where": "this specification: ",
    },
  ],
}
`;

exports[`orgOrGroupTenant path '/groups/{group_id}/thing' is valid: true 1`] = `
Object {
  "base": Object {
    "info": Object {
      "title": "OpenAPI",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {},
  },
  "changelog": Array [],
  "next": Object {
    "info": Object {
      "title": "OpenAPI",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/groups/{group_id}/thing": Object {},
    },
  },
  "results": Array [
    Object {
      "change": Object {
        "location": Object {
          "conceptualLocation": Object {
            "method": "",
            "path": "This Specification",
          },
          "conceptualPath": Array [],
          "jsonPath": "/",
          "kind": "API",
        },
      },
      "condition": "have an org or group tenant",
      "docsLink": "https://github.com/snyk/sweater-comb/blob/main/docs/standards.md#organization-and-group-tenants-for-resources",
      "effectiveOnDate": 2021-07-01T00:00:00.000Z,
      "isMust": true,
      "isShould": false,
      "passed": true,
      "where": "this specification: ",
    },
  ],
}
`;

exports[`orgOrGroupTenant path '/groups/thing' is valid: false 1`] = `
Object {
  "base": Object {
    "info": Object {
      "title": "OpenAPI",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {},
  },
  "changelog": Array [],
  "next": Object {
    "info": Object {
      "title": "OpenAPI",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/groups/thing": Object {},
    },
  },
  "results": Array [
    Object {
      "change": Object {
        "location": Object {
          "conceptualLocation": Object {
            "method": "",
            "path": "This Specification",
          },
          "conceptualPath": Array [],
          "jsonPath": "/",
          "kind": "API",
        },
      },
      "condition": "have an org or group tenant",
      "docsLink": "https://github.com/snyk/sweater-comb/blob/main/docs/standards.md#organization-and-group-tenants-for-resources",
      "effectiveOnDate": 2021-07-01T00:00:00.000Z,
      "error": "expected support for org or group tenant: expected [] to have a length of 1 but got +0",
      "isMust": true,
      "isShould": false,
      "passed": false,
      "where": "this specification: ",
    },
  ],
}
`;

exports[`orgOrGroupTenant path '/org/{org_id}' is valid: false 1`] = `
Object {
  "base": Object {
    "info": Object {
      "title": "OpenAPI",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {},
  },
  "changelog": Array [],
  "next": Object {
    "info": Object {
      "title": "OpenAPI",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/org/{org_id}": Object {},
    },
  },
  "results": Array [
    Object {
      "change": Object {
        "location": Object {
          "conceptualLocation": Object {
            "method": "",
            "path": "This Specification",
          },
          "conceptualPath": Array [],
          "jsonPath": "/",
          "kind": "API",
        },
      },
      "condition": "have an org or group tenant",
      "docsLink": "https://github.com/snyk/sweater-comb/blob/main/docs/standards.md#organization-and-group-tenants-for-resources",
      "effectiveOnDate": 2021-07-01T00:00:00.000Z,
      "error": "expected support for org or group tenant: expected [] to have a length of 1 but got +0",
      "isMust": true,
      "isShould": false,
      "passed": false,
      "where": "this specification: ",
    },
  ],
}
`;

exports[`orgOrGroupTenant path '/orgs/{org_id}/thing' is valid: true 1`] = `
Object {
  "base": Object {
    "info": Object {
      "title": "OpenAPI",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {},
  },
  "changelog": Array [],
  "next": Object {
    "info": Object {
      "title": "OpenAPI",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/orgs/{org_id}/thing": Object {},
    },
  },
  "results": Array [
    Object {
      "change": Object {
        "location": Object {
          "conceptualLocation": Object {
            "method": "",
            "path": "This Specification",
          },
          "conceptualPath": Array [],
          "jsonPath": "/",
          "kind": "API",
        },
      },
      "condition": "have an org or group tenant",
      "docsLink": "https://github.com/snyk/sweater-comb/blob/main/docs/standards.md#organization-and-group-tenants-for-resources",
      "effectiveOnDate": 2021-07-01T00:00:00.000Z,
      "isMust": true,
      "isShould": false,
      "passed": true,
      "where": "this specification: ",
    },
  ],
}
`;

exports[`orgOrGroupTenant path '/orgs/thing' is valid: false 1`] = `
Object {
  "base": Object {
    "info": Object {
      "title": "OpenAPI",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {},
  },
  "changelog": Array [],
  "next": Object {
    "info": Object {
      "title": "OpenAPI",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/orgs/thing": Object {},
    },
  },
  "results": Array [
    Object {
      "change": Object {
        "location": Object {
          "conceptualLocation": Object {
            "method": "",
            "path": "This Specification",
          },
          "conceptualPath": Array [],
          "jsonPath": "/",
          "kind": "API",
        },
      },
      "condition": "have an org or group tenant",
      "docsLink": "https://github.com/snyk/sweater-comb/blob/main/docs/standards.md#organization-and-group-tenants-for-resources",
      "effectiveOnDate": 2021-07-01T00:00:00.000Z,
      "error": "expected support for org or group tenant: expected [] to have a length of 1 but got +0",
      "isMust": true,
      "isShould": false,
      "passed": false,
      "where": "this specification: ",
    },
  ],
}
`;

exports[`orgOrGroupTenant path '/thing' is valid: false 1`] = `
Object {
  "base": Object {
    "info": Object {
      "title": "OpenAPI",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {},
  },
  "changelog": Array [],
  "next": Object {
    "info": Object {
      "title": "OpenAPI",
      "version": "0.0.0",
    },
    "openapi": "3.0.1",
    "paths": Object {
      "/thing": Object {},
    },
  },
  "results": Array [
    Object {
      "change": Object {
        "location": Object {
          "conceptualLocation": Object {
            "method": "",
            "path": "This Specification",
          },
          "conceptualPath": Array [],
          "jsonPath": "/",
          "kind": "API",
        },
      },
      "condition": "have an org or group tenant",
      "docsLink": "https://github.com/snyk/sweater-comb/blob/main/docs/standards.md#organization-and-group-tenants-for-resources",
      "effectiveOnDate": 2021-07-01T00:00:00.000Z,
      "error": "expected support for org or group tenant: expected [] to have a length of 1 but got +0",
      "isMust": true,
      "isShould": false,
      "passed": false,
      "where": "this specification: ",
    },
  ],
}
`;
